modules/renter/hostdb/hostdb.go

```diff
@@ -52,31 +51,27 @@ type HostDB struct {
 
 	// The hostTree is the root node of the tree that organizes hosts by
 	// weight. The tree is necessary for selecting weighted hosts at
-	// random. 'activeHosts' provides a lookup from hostname to the the
-	// corresponding node, as the hostTree is unsorted. A host is active if
-	// it is currently responding to queries about price and other
-	// settings.
-	hostTree    *hosttree.HostTree
-	activeHosts map[modules.NetAddress]*hostEntry
-
-	// allHosts is a simple list of all known hosts by their network address,
-	// including hosts that are currently offline.
-	allHosts map[modules.NetAddress]*hostEntry
+	// random.
+	hostTree *hosttree.HostTree
 
 	// the scanPool is a set of hosts that need to be scanned. There are a
 	// handful of goroutines constantly waiting on the channel for hosts to
 	// scan.
-	scanList []*hostEntry
-	scanPool chan *hostEntry
+	scanList []modules.HostDBEntry
+	scanPool chan modules.HostDBEntry
 	scanWait bool
+	online   bool
 
 	blockHeight types.BlockHeight
 	lastChange  modules.ConsensusChangeID
 }
 
 // New returns a new HostDB.
-func New(cs consensusSet, persistDir string) (*HostDB, error) {
+func New(g modules.Gateway, cs consensusSet, persistDir string) (*HostDB, error) {
```

Comment:
Why do you say we want the entire gateway? We only want to determine whether we are online. The downside of using a full gateway is that now you need to supply a full gateway whenever you want to construct a HostDB for testing. You can't easily mock the gateway in order to test `threadedOnlineCheck`.

I agree that it shouldn't have a `Peers` method though. The fact that we use the gateway's peers to determine online status is an implementation detail. What you really want is a `HasInternetConnection() bool` method. That makes it clear what your intention is. Then wrap the full gateway in a type that uses `g.Peers()` to implement that method.

I know that restrictive typing is a pain, but it pays off in the long run by minimizing surface area and clarifying intent. Think of it as a form of documentation.